[JENKINS-35238] Parent POM and PCT fix.

armfergom opened the pull request:
- [x] Upgrade to parent pom 2.9 
- [x] Fix PCT against all baselines from 1.565.3 to 2.7
- [x] Fix Javadoc issues

There are some changes in this PR that need some clarifications. Please, see them in inline comments.

@reviewbybees esp. @oleg-nenashev

armfergom commented on docker-traceability-api/pom.xml:
```diff
@@ -132,6 +73,20 @@
+        <exclusion>
+          <groupId>org.apache.httpcomponents</groupId>
+          <artifactId>httpmime</artifactId>
+        </exclusion>
```
Exclude `httpclient` and `httpmime` since `docker-java` brings versions that are not compatible with HtmlUnit when run against 2.7 baseline.

armfergom commented on docker-traceability-api/pom.xml:
```diff
@@ -132,6 +73,20 @@
+          <groupId>org.glassfish.hk2</groupId>
+          <artifactId>hk2-utils</artifactId>
+        </exclusion>
+      </exclusions>
```
This is a tricky one. This dependency, by using the `maven-dependency-plugin` [includes the sources](https://github.com/hk2-project/hk2/blob/master/hk2-utils/pom.xml#L100) of `org.jvnet.tiger-types`. On the other hand, the core brings its own dependency of `org.jvnet.tiger-types` (transitive dependency of stapler)

From baseline 2.2 where stapler version is bumped to [1.243](https://github.com/jenkinsci/jenkins/blob/53f665c4539c270b8745481cc9380c28b58c8a1b/core/pom.xml#L42) (and transitively `tiger-types` to 2.2), the PCT fails since stapler calls a method that is not present in the version of tiger types that `hk2-utils` includes in its source code, so this dependency needs to be excluded so that the right version of `tiger-types` is used.

KostyaSha replied:
right for whom?

armfergom replied:
I am not sure I understand your question @KostyaSha

KostyaSha replied:
You need compare library version and library usage to understand whether it save to replacements switches. 
But keeping in mind that this plugin has totally wrong dependency coming from docker-java+jenkins you can just merge and nobody will complain even if you broke it :)

armfergom commented on docker-traceability-plugin/src/main/java/org/jenkinsci/plugins/docker/traceability/JenkinsInstance.java:
```diff
@@ -0,0 +1,29 @@
+    public static Jenkins get() {
+        return Jenkins.getInstance();
+    }
+}
```
This is kind of hacky and I don't like it, but it was the alternative I could think of without polluting all methods using `Jenkins.getInstance()` with `@SuppressFBWarnings` which would increase the technical debt even more than this IMHO.

I did not consider ignoring FindBugs errors at all, I think the check should be performed and this is no reason to disable it with `<findbugs.failOnError>false</findbugs.failOnError>`

KostyaSha commented on docker-traceability-plugin/src/main/java/org/jenkinsci/plugins/docker/traceability/JenkinsInstance.java:
```diff
@@ -0,0 +1,29 @@
+
+import jenkins.model.Jenkins;
+
+public class JenkinsInstance {
```
Should be named as Utils. Must have private constructor.

armfergom replied:
Agreed, must have a `private` constructor but seems to me that `Utils` name is too generic given that the utility class only wraps the call to `Jenkins.getInstance()`.

May I suggest `JenkinsInstanceWrapper`?

KostyaSha replied:
That only single static method. JenkinsUtils/Helper i.e.

oleg-nenashev replied:
This class must be also restricted

oleg-nenashev commented:
Still need to review it

oleg-nenashev commented on docker-traceability-api/pom.xml:
```diff
@@ -94,36 +65,6 @@
         </configuration>
       </plugin>
-      <plugin>
-        <artifactId>maven-enforcer-plugin</artifactId>
```
From what I see these enforcer settings do not exist in the parent POM

armfergom replied:
Very true, the enforcer settings in the parent pom are not the same. Will fix this.

oleg-nenashev commented on docker-traceability-plugin/src/main/java/org/jenkinsci/plugins/docker/traceability/JenkinsInstance.java:
```diff
@@ -0,0 +1,29 @@
+     * So, until the baseline is updated to 2.4, there is no way to make the build
+     * pass for all baselines due to FindBugs errors (unless ignored), i.e:
+     *      - Cannot remove the null check because it fails with baselines 1.565.1 - 2.3
+     *      - Cannot use the null check because it fails with baselines 2.4++
```
It should not fail. It should be just a warning in FindBugs

armfergom replied:
It does fail

oleg-nenashev commented:
🐛 for the non-restricted Util class. Honestly I also see no reason for introducing such hack just to fix the build for 2.4. There is no short-term plans to migrate to this version. What is the purpose of such baseline support?

armfergom commented:
@oleg-nenashev will address the non-restricted Util class. 

The reason to have this is to be able to run the PCT against all versions, including 2.4++.

oleg-nenashev commented:
@armfergom 
PCT should not fail on static analysis warnings. Otherwise it will be a never-ending story. It should only print warnings, so the best way would be to override Maven settings in PCT

armfergom commented:
Closing and opening to retrigger CI job

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev commented:
@armfergom Just to clarify, I'll be fine with this particular PR if it contains restricted definition

armfergom commented:
@oleg-nenashev I have addressed your comment. You are right, the PCT does not fail because of findbugs, so I have reverted that ugly hack.

Build is failing in the CI server although it builds ok on my machine. Looking into it now..

oleg-nenashev commented:
🐛 since tests fail. Not sure why, needs investigation

armfergom commented:
@oleg-nenashev don't ask me why, but build is back to stable. I can tell you what I did though:

1. Triggered build manually in CI served - Failed while checking out, noticed that workspace was empty after the failure (not sure if it was before the failure)
2. Closed and reopened PR again - Worked.

KostyaSha commented:
@armfergom check INFRA issues, jenkins periodically out of space.

armfergom commented:
@KostyaSha @oleg-nenashev I have removed the hk2-utils exclusion since it only starts being a problem from Jenkins v2.2 onwards.

recena commented:
🐝 and 👍

armfergom commented:
@oleg-nenashev @KostyaSha  is this looking better to get it merged now?

kwhetstone commented:
:bee: Looks good!

armfergom commented:
@reviewbybees done

armfergom commented:
@oleg-nenashev would it be possible to get this merged and released? Thanks!

reviewbybees commented:
This pull request has completed our internal processes and we now respectfully request the maintainers of this repository to consider our proposal contained within this pull request for merging.

oleg-nenashev commented on docker-traceability-plugin/pom.xml:
```diff
@@ -127,36 +105,6 @@
-              <rules>
-                <requireJavaVersion>
-                  <version>[1.7,)</version>
-                  <message>Build with Java 1.7 or later, docker-java uses 1.7 APIs and bytecode.</message>
```
I'm not sure it's safe to remove the enforcer here

oleg-nenashev commented:
🐝
